### PR TITLE
Fix broken include of Jinja template from absolute path

### LINKED ISSUE #62611: [BUG] Including Jinja template from absolute path broken
**Description**.

Including Jinja template from absolute path broke with #62043 

Please be as specific as possible and give set-up details.

- [ ] on-prem machine
- [ ] VM (Virtualbox, KVM, etc. please specify)
- [x] VM running on a cloud service, please be explicit and add details
- [ ] container (Kubernetes, Docker, containerd, etc. please specify)
- [ ] or a combination, please be explicit
- [ ] jails if it is FreeBSD

  
**Steps to Reproduce the behavior**
`sudo salt '*' slsutil.renderer salt://jinja/to/render.j2 path_to_include_template=/absolute/path/to/template.j2`

**Expected behavior**
Properly include jinja templates when they are an absolute path. 

**Screenshots**

image

**Versions Report**

```yaml
Salt Version:
          Salt: 3005
 
Dependency Versions:
          cffi: Not Installed
      cherrypy: unknown
      dateutil: 2.7.3
     docker-py: Not Installed
         gitdb: 2.0.6
     gitpython: 3.0.7
        Jinja2: 2.10.1
       libgit2: Not Installed
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 0.6.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: Not Installed
      pycrypto: Not Installed
  pycryptodome: 3.6.1
        pygit2: Not Installed
        Python: 3.8.10 (default, Jun 22 2022, 20:18:18)
  python-gnupg: 0.4.5
        PyYAML: 5.3.1
         PyZMQ: 18.1.1
         smmap: 2.0.5
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.3.2
 
System Versions:
          dist: ubuntu 20.04 focal
        locale: utf-8
       machine: x86_64
       release: 5.15.0-1013-oracle
        system: Linux
       version: Ubuntu 20.04 focal
```
</details>

**Additional context**

The new `if` statement introduced in #62043 fails for absolute paths, and so the old logic that successfully found these in the `for` loop no longer runs:
https://github.com/saltstack/salt/blob/662cfbadfa454c83407216aad1762677348841a1/salt/utils/jinja.py#L200-L203

At this point, there is no early return, and we fall through to the `raise`:
https://github.com/saltstack/salt/blob/662cfbadfa454c83407216aad1762677348841a1/salt/utils/jinja.py#L222

### PR DESCRIPTION
**Pull request by nicholasmhughes**
### What does this PR do?
See issue for details.

### What issues does this PR fix or reference?
Fixes: #62611 

### Previous Behavior
Error when including a template from an absolute path

### New Behavior
Much success

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by Ch3LL** [comment]
would you mind pointing this to the freeze branch so we can get it in for 3005.1?

**Comment by nicholasmhughes** [comment]
done!

File: salt/utils/jinja.py
Comment (s0undt3ch) [inline salt/utils/jinja.py:199]: Wouldn't it be better if we checked it was, in fact, an absolute path?
Just checking for the existence opens up for relative path security issues?

And, I'm currently on PTO so I can't check deeper, but, wouldn't absolute paths also open salt up for security issues?

Comment (nicholasmhughes) [reply salt/utils/jinja.py:199]: Responding from memory, if `_template` ends up in `self.cached`, that means that it's either a relative or absolute path _inside the salt filesystem_. So if it hits the `or` here, then it's _outside_ the salt filesystem. I don't think it's security-relevant whether the path is absolute or relative if we're deciding that we're allowing access outside the salt filesystem (retaining pre-existing functionality). However, I had a similar thought when I was writing the original patch... If we don't merge this PR, the change in functionality could be considered as fixing a security hole.

Does the Core Team want to take a poll and weigh the risk against breaking existing functionality?

**Comment by nicholasmhughes** [comment]
re-run pr-freebsd-131-amd64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-windows-2016-x64-py3-pytest

File: salt/utils/jinja.py
Comment (whytewolf) [reply salt/utils/jinja.py:199]: I would say lets label this as fixing a security hole. not sure rendering outside of the salt fileserver or at least the cache is really needed.

Comment (nicholasmhughes) [reply salt/utils/jinja.py:199]: Looping in @sabedevops and @supfors to speak to the use case.

Comment (nicholasmhughes) [reply salt/utils/jinja.py:199]: also @iaingeorgeson

Comment (sabedevops) [reply salt/utils/jinja.py:199]: To answer the direct question, our use case is that we shard some configuration file templates. Some of the shard reside in an data sources with respect to Salt, and are dropped into place during system provisioning. The template that resides in the salt base serves as an interface to populate required variables and include that external template shard before final rendering.

As an aside, breaking the user experience in this way should always be last resort, in my opinion, and should only happen when the security concern is egregious and cannot be ignored. Fixing a security hole wasn't the goal of the initial patch, for example. Furthermore, I would question what security hole it is really fixing? The types of failures I'm imagining really reside in the process controls implemented by the user outside the scope of Salt. Can you explain what the concern is more concretely and how ensuring the template resides in the salt base addresses that concern? 

Now I'm thinking I have to emergency patch our workflow and the hundreds of salt-masters we deploy. If that is the case, I would really like to understand our exposure in the majority of our systems that precede 3005.

Comment (waynew) [reply salt/utils/jinja.py:199]: I'm not sure that it's actually any kind of security hole. You can already do something like this:

```
/path/to/some/file:
    file.managed:
        - source: /path/to/literally/anywhere/on/the/filesystem
        - template: jinja
```

Unless I'm misunderstanding the path that results in this code being called. If we're talking about the above scenario, there's no security hole to be plugged -- it's just normal Salt functionality.

Comment (waynew) [reply salt/utils/jinja.py:199]: (relative source paths I think also work there, but I'm not checking that right now)

**Comment by nicholasmhughes** [comment]
@twangboy , do you have any ideas on this Windows test failure?

```
File "C:\\Users\\Administrator\\AppData\\Local\\Temp\\kitchen\\testing\\salt\\utils\\jinja.py", line 221, in get_source\n    raise TemplateNotFound(template)\njinja2.exceptions.TemplateNotFound: C:\\Windows\\Temp\\stsuite\\master-AGlZwz\\state-tree\x08ase\test_jinja\\issue-62611.jinja\n\n; line 1\n\n---\n{%- from "C:\\Windows\\Temp\\stsuite\\master-AGlZwz\\state-tree\\base\\test_jinja\\issue-62611.jinja" import myvar with context %}    <======================\n\ntest_jinja/issue-62611/cmd.run:\n  cmd.run:\n    - name: echo MEEP {{ myvar }}\n\n[...]\n---'])
```

The `TemplateNotFound` exception seems to show that there are weird escape sequences in there (`\x08ase` instead of `base`)... but the import line looks like the backslashes are all matched up:

```
{%- from "C:\\Windows\\Temp\\stsuite\\master-AGlZwz\\state-tree\\base\\test_jinja\\issue-62611.jinja" import myvar with context %}
```

Should the path separator be escaped in this string or not?

File: salt/utils/jinja.py
Comment (s0undt3ch) [inline salt/utils/jinja.py:199]: I think the proper check needs to be something like `if _template in self.cached or (os.path.exists(_template) and os.path.isabs(_template)):

Comment (nicholasmhughes) [reply salt/utils/jinja.py:199]: From https://github.com/saltstack/salt/pull/62629#discussion_r964743556 :

> "I don't think it's security-relevant whether the path is absolute or relative if we're deciding that we're allowing access outside the salt filesystem (retaining pre-existing functionality)."

While the title of the PR contains "absolute path" because that's what was reported, the `exists` logic reverts the outcome to the same that was available prior to the patch in 3005.

**Comment by nicholasmhughes** [comment]
re-run pr-centos-7-x86_64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-centos-7-x86_64-py3-tcp-pytest

File: salt/utils/jinja.py
Comment (s0undt3ch) [reply salt/utils/jinja.py:199]: Let's retain the previous functionality.

File: tests/pytests/integration/states/test_file.py
Comment (s0undt3ch) [inline tests/pytests/integration/states/test_file.py:1316]:
```suggestion
    jinja_file = salt_master.state_tree.base.write_path / "{}.jinja".format(name)
```

Comment (s0undt3ch) [reply tests/pytests/integration/states/test_file.py:1316]: This change is not a blocker for merging though.

**Review by s0undt3ch — APPROVED** [review]

**Comment by nicholasmhughes** [comment]
re-run pr-photon-3-x86_64-py3-pytest

**Pull request merged**
